### PR TITLE
MatchTest uses the new Cypher doc build system

### PR DESCRIPTION
**Pull request by henriknyman**
- Completed NewMatchTest and renamed it MatchTest
- Added IDREF to Section in DSL
- Renamed NewPrettyGraphTest to PrettyGraphTest

### CODE REVIEW

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 1]
```diff
@@ -19,347 +19,305 @@
...
+        p("When you know the relationship type you want to match on, you can specify it by using a colon together with the relationship type.")
+        query("MATCH (wallstreet:Movie { title:'Wall Street' })<-[:ACTED_IN]-(actor) RETURN actor.name", assertAllActorsOfWallStreetAreFound) {
+          p("Returns all outgoing relationships from Oliver.")
+          resultTable()
+        }
+      }
+      section("Match by multiple relationship types") {
+        p("To match on one of multiple types, you can specify this by chaining them together with the pipe symbol `|`.")
+        query("MATCH (wallstreet { title:'Wall Street' })<-[:ACTED_IN|:DIRECTED]-(person) RETURN person.name", assertEveryoneConnectedToWallStreetIsFound) {
+          p("Returns nodes with a +ACTED_IN+ or +DIRECTED+ relationship to Wall Street.")
```
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: "a" should be "an"!

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 2]
```diff
@@ -19,347 +19,305 @@
...
+          resultTable()
+        }
+      }
+      section("Match by relationship type and use an identifier") {
+        p("If you both want to introduce an identifier to hold the relationship, and specify the relationship type you want, just add them both, like this:")
+        query("MATCH (wallstreet { title:'Wall Street' })<-[r:ACTED_IN]-(actor) RETURN r", assertRelationshipsToWallStreetAreReturned) {
```
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I'd suggest we return `r.role` instead here.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 3]
```diff
@@ -19,347 +19,305 @@
...
+        }
+      }
+    }
+    section("Relationships in depth") {
+      note {
+        p("Inside a single pattern, relationships will only be matched once. You can read more about this in <<cypherdoc-uniqueness>>")
+      }
+      section("Relationship types with uncommon characters") {
+        p("Sometime your database will have types with non-letter characters, or with spaces in them. Use +`+ (backtick) to quote these.")
+        query("MATCH (n { name:'Rob Reiner' })-[r:`TYPE THAT HAS SPACE IN IT`]->() RETURN count(*)", assertTypeWithRelsExist) {
```
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Can we avoid hitting keywords (has, in) in the reltype? The syntax highlighter isn't very smart.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 4]
```diff
@@ -19,347 +19,305 @@
...
+      }
+      section("Multiple relationships") {
+        p("Relationships can be expressed by using multiple statements in the form of `()--()`, or they can be strung together, like this:")
+        query("match (charlie {name:'Charlie Sheen'})-[:ACTED_IN]->(movie)<-[:DIRECTED]-(director) return movie.title, director.name", assertFindAllDirectors) {
+          p("Returns the three nodes in the path.")
+          resultTable()
+        }
+      }
+      section("Variable length relationships") {
+        p("Nodes that are a variable number of relationship->node hops away can be found using the following syntax: `-[:TYPE*minHops..maxHops]->`. ")
```
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: This was lost: `minHops and maxHops are optional and default to 1 and infinity respectively. When no bounds are given the dots may be omitted.`

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 5]
```diff
@@ -19,347 +19,305 @@
...
+        query(
+          """MATCH (martin:Person {name:"Martin Sheen"} ),
+            |      (oliver:Person {name:"Oliver Stone"}),
+            |      p = shortestPath( (martin)-[*..15]-(oliver) )
+            |RETURN p""", assertShortestPathLength) {
+          resultTable()
+          p(
+            """This means: find a single shortest path between two nodes, as long as the path is max 15 relationships long. Inside of the parentheses
+              | you define a single link of a path -- the starting node, the connecting relationship and the end node. Characteristics describing the relationship
+              | like relationship type, max hops and direction are all used when finding the shortest path. You can also mark the path as optional.""")
```
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: `You can also mark the path as optional.` isn't very clear.

**Comment by nawroth** [comment]
Some sections are gone: Zero length paths, Named path, Matching on a bound relationship. What's up with that?

Great that you're returning properties instead of full nodes/rels in some places!

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 6]
```diff
@@ -19,347 +19,305 @@
...
-        val allNodes: List[Node] = GlobalGraphOperations.at(db).getAllNodes.asScala.toList
-
-        assertEquals(allNodes, p.columnAs[Node]("a").toList)
+    synopsis("The `MATCH` clause is used to search for the pattern described in it.")
+    section("Introduction") {
+      p( """The `MATCH` clause allows you to specify the patterns Neo4j will search for in the database.
+           |This is the primary way of getting data into the current set of bindings.
+           |It is worth reading up more on the specification of the patterns themselves in <<introduction-pattern>>.""")
+      p( """`MATCH` is often coupled to a `WHERE` part which adds restrictions, or predicates, to the `MATCH` patterns, making them more specific.
+           |The predicates are part of the pattern description, not a filter applied after the matching is done.
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Is this true? I know for sure it's implemented as a filter after the match, except for inlining some property predicates into an indexseek.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 7]
```diff
@@ -19,347 +19,305 @@
...
+           |_This means that `WHERE` should always be put together with the `MATCH` clause it belongs to._""")
+      p( """`MATCH` can occur at the beginning of the query or later, possibly after a `WITH`.
+           |If it is the first clause, nothing will have been bound yet, and Neo4j will design a search to find the results matching the clause and any associated predicates specified in any `WHERE` part.
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Do we have some convention on whether to use the name `Neo4j` or the name `Cypher` in situations like these?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 8]
```diff
@@ -19,347 +19,305 @@
...
+           |This could involve a scan of the database, a search for nodes of a certain label, or a search of an index to find starting points for the pattern matching.
+           |Nodes and relationships found by this search are available as _bound pattern elements,_ and can be used for pattern matching of sub-graphs.
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:61]: Nitpick: have the comma after the italics.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 5 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: This does not make sense to me either, and a review of the shortestPath code does not clarify it for me. I think I'll delete this line and add instead a comment about the new predicate-pull-in capability.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 9]
```diff
@@ -19,347 +19,305 @@
...
+           |They can also be used in any further `MATCH` clauses, where Neo4j will use the known elements, and from there find further unknown elements.""")
+      p( """Cypher is declarative, and so usually the query itself does not specify the algorithm to use to perform the search.
+           |Neo4j will automatically work out the best approach to finding start nodes and matching patterns.
+           |Predicates in `WHERE` parts can be evaluated before pattern matching, during pattern matching, or after finding matches.
+           |However, there are cases where you can influence the decisions taken by the query compiler.
+           |Read more about indexes in <<query-schema-index>>, and more about the specifying index hints to force Neo4j to use a specific index in <<query-using>>.""")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Perhaps mention that the `USING` page covers more than just index hints? Eg, scan hints, join hints.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 10]
```diff
@@ -19,347 +19,305 @@
...
-      assertions = (p) => assertEquals(2, p.size)
-    )
-  }
+      p("The following graph is used for the examples below:")
+      graphViz()
+    }
+    section("Basic node finding") {
+      section("Get all nodes") {
+        p("By just specifying a pattern with a single node and no labels, all nodes in the graph will be returned.")
+        query("MATCH (n) RETURN n", assertAllNodesReturned) {
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:77]: We don't want to suggest returning the full node. What about `RETURN id(n)`?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 11]
```diff
@@ -19,347 +19,305 @@
...
+      }
+      section("Get all nodes with a label") {
+        p("Getting all nodes with a label on them is done with a single node pattern where the node has a label on it.")
+        query("MATCH (movie:Movie) RETURN movie.title", assertAllMoviesAreReturned) {
+          p("Returns all the movies in the database.")
+          resultTable()
+        }
+      }
+      section("Related nodes") {
+        p("The symbol `--` means _related to,_ without regard to type or direction of the relationship.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:90]: Another comma nitpick

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 12]
```diff
@@ -19,347 +19,305 @@
...
+        query("MATCH (director { name:'Oliver Stone' })--(movie) RETURN movie.title", assertWallStreetIsReturned) {
+          p("Returns all the movies directed by Oliver Stone.")
+          resultTable()
+        }
+      }
+      section("Match with labels") {
+        p("To constrain your pattern with labels on nodes, you add it to your pattern nodes, using the label syntax.")
+        query("MATCH (:Person { name:'Oliver Stone' })--(movie:Movie) RETURN movie.title", assertWallStreetIsReturned) {
+          p("Return any nodes connected with the +Person+ Oliver that are labeled +Movie+.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: @nawroth What convention do we have for + and ` signs? When to use which etc

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 13]
```diff
@@ -19,347 +19,305 @@
...
+        p("If an identifier is needed, either for filtering on properties of the relationship, or to return the relationship, this is how you introduce the identifier.")
+        query("MATCH (:Person { name:'Oliver Stone' })-[r]->(movie) RETURN type(r)", assertRelationshipIsDirected) {
+          p("Returns all outgoing relationships from Oliver.")
+          resultTable()
+        }
+      }
+      section("Match by relationship type") {
+        p("When you know the relationship type you want to match on, you can specify it by using a colon together with the relationship type.")
+        query("MATCH (wallstreet:Movie { title:'Wall Street' })<-[:ACTED_IN]-(actor) RETURN actor.name", assertAllActorsOfWallStreetAreFound) {
+          p("Returns all outgoing relationships from Oliver.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: This is clearly wrong. It should say something like `Returns all actors that acted in the movie Wall Street.`

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 14]
```diff
@@ -19,347 +19,305 @@
...
+        p("If you both want to introduce an identifier to hold the relationship, and specify the relationship type you want, just add them both, like this:")
+        query("MATCH (wallstreet { title:'Wall Street' })<-[r:ACTED_IN]-(actor) RETURN r", assertRelationshipsToWallStreetAreReturned) {
+          p("Returns nodes that +ACTED_IN+ Wall Street.")
+          resultTable()
+        }
+      }
+    }
+    section("Relationships in depth") {
+      note {
+        p("Inside a single pattern, relationships will only be matched once. You can read more about this in <<cypherdoc-uniqueness>>")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: This paragraph is missing a full stop.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 15]
```diff
@@ -19,347 +19,305 @@
...
+      }
+      section("Relationship types with uncommon characters") {
+        p("Sometime your database will have types with non-letter characters, or with spaces in them. Use +`+ (backtick) to quote these.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Sometime`s`

Also I would prefer to say `escape these` over `quote these`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 2 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: +1

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 16]
```diff
@@ -19,347 +19,305 @@
...
+        query("MATCH (n { name:'Rob Reiner' })-[r:`TYPE THAT HAS SPACE IN IT`]->() RETURN count(*)", assertTypeWithRelsExist) {
+          resultTable()
+        }
+      }
+      section("Multiple relationships") {
+        p("Relationships can be expressed by using multiple statements in the form of `()--()`, or they can be strung together, like this:")
+        query("match (charlie {name:'Charlie Sheen'})-[:ACTED_IN]->(movie)<-[:DIRECTED]-(director) return movie.title, director.name", assertFindAllDirectors) {
+          p("Returns the three nodes in the path.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Incorrect. It returns movie titles and director names.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 17]
```diff
@@ -19,347 +19,305 @@
...
+          resultTable()
+        }
+      }
+      section("Variable length relationships") {
+        p("Nodes that are a variable number of relationship->node hops away can be found using the following syntax: `-[:TYPE*minHops..maxHops]->`. ")
+        query("match (martin {name:'Charlie Sheen'})-[:ACTED_IN*1..3]-(movie:Movie) return movie.title", assertAllMoviesAreReturned) {
+          p("Returns the three nodes in the path.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Incorrect. It returns movie titles.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 18]
```diff
@@ -19,347 +19,305 @@
...
+          resultTable()
+        }
+      }
+      section("Relationship identifier in variable length relationships") {
+        p("When the connection between two nodes is of variable length, a relationship identifier becomes an collection of relationships.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: `a` collection, not `an`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 19]
```diff
@@ -19,347 +19,305 @@
...
+        query("MATCH (actor { name:'Charlie Sheen' })-[r:ACTED_IN*2]-(co_actor)\nRETURN r", assertActedInRelationshipsAreReturned) {
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Remove `\n`, and prefer `RETURN r.role`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 20]
```diff
@@ -19,347 +19,305 @@
...
+          p("The query returns a collection of relationships.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: If the return is changed like my suggestion above, also change this.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 21]
```diff
@@ -19,347 +19,305 @@
...
+        }
+      }
+      section("Match with properties on a variable length path") {
+        val initQuery =
+          """MATCH (charlie:Person {name:'Charlie Sheen'}), (martin:Person {name:'Martin Sheen'})
+            |CREATE (charlie)-[:X {blocked:false}]->(:Unblocked)<-[:X {blocked:false}]-(martin)
+            |CREATE (charlie)-[:X {blocked:true}]->(:Blocked)<-[:X {blocked:false}]-(martin)"""
+        p("""A variable length relationship with properties defined on in it means that all
+            |relationships in the path must have the property set to the given value. In this query,
+            |there are two paths between Charlie Sheen and his dad Martin Sheen. One of the includes a
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: `the` -> `them`, also I prefer `father` over `dad`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 22]
```diff
@@ -19,347 +19,305 @@
...
+            |``blocked'' relationship and the other doesn't. In this case we first alter the original
+            |graph by using the following query to add ``blocked'' and ``unblocked'' relationships:""")
+        query(initQuery, assertBlockingRelationshipsAdded) {
+          p("This means that we are starting out with the following graph: ")
+          graphViz()
+        }
+        query("MATCH p = (charlie:Person)-[* {blocked:false}]-(martin:Person) " +
+          "WHERE charlie.name = 'Charlie Sheen' AND martin.name = 'Martin Sheen' " +
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Prefer a multiline string here.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 23]
```diff
@@ -19,347 +19,305 @@
...
+      section("Single shortest path") {
+        p("Finding a single shortest path between two nodes is as easy as using the `shortestPath` function. It's done like this:")
+        query(
+          """MATCH (martin:Person {name:"Martin Sheen"} ),
+            |      (oliver:Person {name:"Oliver Stone"}),
+            |      p = shortestPath( (martin)-[*..15]-(oliver) )
+            |RETURN p""", assertShortestPathLength) {
+          resultTable()
+          p(
+            """This means: find a single shortest path between two nodes, as long as the path is max 15 relationships long. Inside of the parentheses
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Unnecessary line break.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 24]
```diff
@@ -19,347 +19,305 @@
...
+              | you define a single link of a path -- the starting node, the connecting relationship and the end node. Characteristics describing the relationship
+              | like relationship type, max hops and direction are all used when finding the shortest path. You can also mark the path as optional.""")
+        }
+      }
+      section("All shortest paths") {
+        p("Finds all the shortest paths between two nodes.")
+        query("""match (martin:Person {name:"Martin Sheen"} ),
+                |      (michael:Person {name:"Michael Douglas"}),
+                |      p = allShortestPaths( (martin)-[*]-(michael) ) return p""", assertAllShortestPaths) {
+          p("Finds the two shortest paths between Martin and Michael.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:258]: Misleading. There might be only two shortest paths, but this should in general return any number of them.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 25]
```diff
@@ -19,347 +19,305 @@
...
+          resultTable()
+        }
+      }
+    }
+    section("Get node or relationship by id") {
+      section("Node by id", "match-node-by-id") {
+        p("Search for nodes by id can be done with the 'id' function in a predicate.")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Search`ing` for nodes [...]

Also, what convention do we have for functions? IMO, we should use backticks and empty parenthesis, like this `id()`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 26]
```diff
@@ -19,347 +19,305 @@
...
+        note {
+          p("""Neo4j reuses its internal ids when nodes and relationships are deleted.
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Is this true? I thought we _didn't_ reuse ids?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 27]
```diff
@@ -19,347 +19,305 @@
...
+              |This means that applications using, and relying on internal Neo4j ids, are brittle or at risk of making mistakes.
+              |Rather use application generated ids.""")
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I'd reword this to something like `For this reason, we recommend that a client uses application generated ids.`

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 28]
```diff
@@ -19,347 +19,305 @@
...
+        }
+        query("match (n) where id(n) = 0 return n", assertHasNodes(0)) {
+          p("The corresponding node is returned.")
+          resultTable()
+        }
+      }
+      section("Relationship by id") {
+        p("""
+            |Search for relationships by id can be done with the 'id' function in a predicate.
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I'd prefer `id()` here.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala [hunk 29]
```diff
@@ -19,347 +19,305 @@
...
+          resultTable()
+        }
+      }
+    }
+  }.build()
+
+  private def assertAllNodesReturned = ResultAndDbAssertions((p, db) => {
+    val tx = db.beginTx()
+    try {
+      val allNodes: List[Node] = GlobalGraphOperations.at(db).getAllNodes.asScala.toList
```
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:300]: I guess it doesn't matter much, but I recall @pontusmelke saying `toList` is much slower than `toVector`. How large is this db?

**Comment by Mats-SX** [comment]
Sorry for being so 'thorough' in this review; it just feels like the first parts of the manual sets the bar for the rest of it, meaning that if we manage to keep everything tidy and using good conventions now, we might be able to keep doing so for some time.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 7 above)
Comment (systay) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: Yes. Cypher is a language and never an active participant. Neo4j does things, and Cypher is the language used to express what should be done

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 26 above)
Comment (systay) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: We do reuse them in a lot of situations. Not always though - HA makes it much more difficult. Maybe that's what you are thinking about?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 8 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:61]: That's true for the semantics, but for the look, it might be better as it stands. The difference is of course much more visible when it comes to characters like `!?;:`.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 10 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:77]: Please not `id()` !

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 25 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: +1

**Comment by nawroth** [comment]
@Mats-SX That's the spirit!

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 8 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:61]: Fair enough.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 10 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:77]: Then what about `count(n)` or `count(*)`?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 26 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: No, I was just misinformed.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 6 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I believe the intention of this sentence is to emphasise the coupling of MATCH and WHERE, because the exact application of the WHERE depends on the logical plan, not on the Cypher syntax. So perhaps the WHERE is used in the leaf plan, perhaps in a post match filter, but that is a question of plan optimization, not of the Cypher syntax. I suggest we keep this section the same for now, and only deal with these subtleties in the section on query plan descriptions. Should we make a link to that section here?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 9 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I think the phrase 'index hints' is general enough to cover scan hints (but not join hints). Since we are not pushing join hints as a feature people should jump on, this text seems ok to me.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 10 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:77]: MATCH (n) RETURN count(n)
Such a classic query. I'd feel like I was spoiling something by changing this. I know I use that query a lot in smaller databases, especially in the browser.

The policy I used when replacing returning nodes with something else we usually to replace in all cases where the apparent intention of the example was not specifically about returning nodes. In this case that seemed to be the primary intention, so it seemed reasonable to leave alone.

I also think that the idea about not returning nodes is not that we think returning nodes is always a bad idea, only that people probably do it far more than they should, and do it in cases when it is far from optimal, or even not what they actually were actually after. So all queries that could be written without returning nodes and still maintain the same intent, should be, and the rest not.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 24 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:258]: I believe the purpose of these paragraphs directly before the results table are to provide a textual description of exactly the results provided, and can be quite specific to the data and query itself. In this case, I actually prefer this text.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 6 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: But this wording explicitly states that the predicates _are not_ a filter, when they very well _may be_. Perhaps we could reword it to `The predicates are part of the pattern description, and should not be considered a filter [...]`?

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 9 above)
Comment (Mats-SX) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: You think scan hints is included in index hints? I disagree. Indexes are label/property pairs, not just a label.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 29 above)
Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:300]: Small db (7 nodes). I think we can live with this one :-)

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 9 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala]: I'd rather like it to be less specific along the lines of `Read more about indexes in <<query-schema-index>>, and more about specifying hints to force Neo4j to solve a query in a specific way in <<query-using>>.`

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 10 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:77]: This isn't easy. As this is the first example on the page, it should ideally be both a super simple query _and_ very clean output.

As the next example does return a property and not full nodes I think we can keep this one as it is.

File: manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala (hunk 24 above)
Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:258]: We use to add those paragraphs _after_ the query. Anyhow, just make it clear that the text is about the example query. `The following example query finds` or similar.

Comment (craigtaverner) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:258]: All the other examples have text starting with 'Returns ....' or 'Finds ....'. I was going for consistency, but I think we do not have that between pages either. Perhaps a review of this concept manual-wide is worth considering.

Comment (nawroth) [inline manual/cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/MatchTest.scala:258]: That's partially an effect of the confusing old tooling. And it gets worse when the sentence is put before the query. I suggest we keep an eye on small improvements like this when we move content to the new tests. A wider review will be more approachable after we have moved all content to the new system!
